### PR TITLE
[REVIEW] Add a double type dispatcher.

### PR DESCRIPTION
**Pull request by bdice**
This PR adds a double type dispatcher, for use in the AST (#5493, #5494).

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

File: cpp/include/cudf/utilities/type_dispatcher.hpp [hunk 1]
```diff
@@ -419,5 +419,53 @@ CUDA_HOST_DEVICE_CALLABLE constexpr decltype(auto) type_dispatcher(cudf::data_ty
   }
 }
 
+namespace {
```
Comment (jrhemstad) [inline cpp/include/cudf/utilities/type_dispatcher.hpp:422]: Don't use an anonymous namespace in a header.

http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rs-unnamed

File: cpp/include/cudf/utilities/type_dispatcher.hpp [hunk 2]
```diff
@@ -419,5 +419,53 @@ CUDA_HOST_DEVICE_CALLABLE constexpr decltype(auto) type_dispatcher(cudf::data_ty
...
+#pragma nv_exec_check_disable
+  template <typename T2, typename F, typename... Ts>
+  CUDA_HOST_DEVICE_CALLABLE decltype(auto) operator()(F&& f, Ts&&... args) const noexcept
+  {
+    return f.template operator()<T1, T2>(std::forward<Ts>(args)...);
+  }
+};
+
+template <template <cudf::type_id> typename IdTypeMap>
+struct dispatch_second_type {
```
Comment (jrhemstad) [inline cpp/include/cudf/utilities/type_dispatcher.hpp:434]: I think naming would be more intuitive to call this one `dispatch_first_type`, and the other one `dispatch_second_type`.

File: cpp/include/cudf/utilities/type_dispatcher.hpp [hunk 3]
```diff
@@ -419,5 +419,53 @@ CUDA_HOST_DEVICE_CALLABLE constexpr decltype(auto) type_dispatcher(cudf::data_ty
...
+#pragma nv_exec_check_disable
+  template <typename T1, typename F, typename... Ts>
+  CUDA_HOST_DEVICE_CALLABLE decltype(auto) operator()(cudf::data_type type2,
+                                                      F&& f,
+                                                      Ts&&... args) const noexcept
```
Comment (jrhemstad) [inline cpp/include/cudf/utilities/type_dispatcher.hpp:439]: If this function is going to be `noexcept`, it has to be conditional upon if the callable is `noexcept`.

File: cpp/include/cudf/utilities/type_dispatcher.hpp [hunk 4]
```diff
@@ -419,5 +419,53 @@ CUDA_HOST_DEVICE_CALLABLE constexpr decltype(auto) type_dispatcher(cudf::data_ty
   }
 }
 
+namespace {
+template <typename T1>
+struct dispatch_second_type_impl {
+#pragma nv_exec_check_disable
+  template <typename T2, typename F, typename... Ts>
+  CUDA_HOST_DEVICE_CALLABLE decltype(auto) operator()(F&& f, Ts&&... args) const noexcept
```
Comment (jrhemstad) [inline cpp/include/cudf/utilities/type_dispatcher.hpp:427]: Same here.

File: cpp/include/cudf/utilities/type_dispatcher.hpp (hunk 3 above)
Comment (bdice) [reply cpp/include/cudf/utilities/type_dispatcher.hpp:439]: That was copied from the aggregation dispatcher -- it probably makes sense to remove the `noexcept` here.

File: cpp/include/cudf/utilities/type_dispatcher.hpp [hunk 5]
```diff
@@ -419,5 +419,56 @@ CUDA_HOST_DEVICE_CALLABLE constexpr decltype(auto) type_dispatcher(cudf::data_ty
...
+ *
+ * @param type1 The `data_type` used to dispatch a type for the first template
+ * parameter of the callable `F`
+ * @param type2 The `data_type` used to dispatch a type for the second template
+ * parameter of the callable `F`
+ * @param args Parameter pack forwarded to the `operator()` invocation `F`.
+ */
+#pragma nv_exec_check_disable
+template <template <cudf::type_id> typename IdTypeMap = id_to_type_impl, typename F, typename... Ts>
+CUDA_HOST_DEVICE_CALLABLE constexpr decltype(auto) type_double_dispatcher(cudf::data_type type1,
```
Comment (jrhemstad) [inline cpp/include/cudf/utilities/type_dispatcher.hpp:461]: I think `double_type_dispatcher` would be more appropriate.

Comment (bdice) [reply cpp/include/cudf/utilities/type_dispatcher.hpp]: Resolved in d58245ae942c1c34882f30b0d4c3dba904031c21.

File: cpp/tests/types/type_dispatcher_test.cu [hunk 6]
```diff
@@ -84,4 +84,95 @@ TEST_P(IdDispatcherTest, IdToType)
   EXPECT_TRUE(cudf::type_dispatcher(cudf::data_type{t}, verify_dispatched_type{}, t));
 }
 
+template <typename T>
+struct DoubleTypedDispatcherTest : public DispatcherTest {
```
Comment (harrism) [inline cpp/tests/types/type_dispatcher_test.cu:88]:
```suggestion
struct DoubleTypeDispatcherTest : public DispatcherTest {
```

Comment (bdice) [reply cpp/tests/types/type_dispatcher_test.cu:88]: @harrism I was matching the name of the existing `TypedDispatcherTest` but I agree this is a better name. May I change the single-dispatch test names as well?

Comment (harrism) [reply cpp/tests/types/type_dispatcher_test.cu:88]: It's not a big deal. Up to you.

Comment (jrhemstad) [reply cpp/tests/types/type_dispatcher_test.cu:88]: Don't change the names. TypedDispatcherTest reflects the fact that it is the _typed_ fixture class for the type-parameterized tests.

Comment (jrhemstad) [reply cpp/tests/types/type_dispatcher_test.cu:88]: As such, the extension of naming for the double dispatcher would be `TypedDoubleDispatcherTest`.

Comment (bdice) [reply cpp/tests/types/type_dispatcher_test.cu]: Good catch @jrhemstad. Fixed in 748124b7fe786fda6ecacf5d1b4beebc98271ef9.

**Pull request merged**